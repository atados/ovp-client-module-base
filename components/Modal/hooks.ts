import { useMemo, useContext, useRef, useEffect } from 'react'
import nanoid from 'nanoid'
import {
  ModalContext,
  ModalConfig,
  ModalContextType,
} from '~/components/Modal/context'
import { dev } from '~/common/constants'

export interface UseModalOptions<Props>
  extends Partial<Omit<ModalConfig<Props>, 'component' | 'props'>> {
  component: React.ComponentType<Props>
  /**
   * @deprecated
   */
  componentProps?: any
  /**
   * @deprecated
   */
  cardClassName?: string
  onClosePropName?: string
}

export interface UseModalResult<Props> {
  (props?: Props): string
  id: string
  open: (props?: Props) => string
  close: () => void
}

export const useModal = <Props>({
  id: givenId,
  component,
  label,
  className,
  cardClassName: deprecatedCardClassName,
  componentProps: deprectedComponentProps,
  onClosePropName: deprecatedOnClosePropName,
}: UseModalOptions<Props>): UseModalResult<Props> => {
  const modals = useModals()
  const id = useMemo(() => givenId || nanoid(), [givenId])

  if (dev && deprectedComponentProps) {
    console.warn(
      'useModal with componentProps will be deprecated. Pass props at the open call',
    )
  }

  useEffect(() => {
    const lastModal = modals.get(id)
    if (lastModal) {
      modals.replace(id, {
        ...lastModal,
        component,
      })
    }
  }, [component])

  return useMemo(() => {
    const modal: UseModalResult<Props> = ((props?: Props) => {
      if (dev && !props) {
        console.warn(
          `Modal with id ${id} was open without passing props. This will be deprecated`,
        )
      }

      const passedProps = deprectedComponentProps
        ? { ...deprectedComponentProps, ...props }
        : { ...props }

      if (deprecatedOnClosePropName) {
        if (dev) {
          console.warn(
            `onClosePropName option on useModal will be deprecated. Pass it as a prop with \`modal.open({ onClose: () => modal.close() })\``,
          )
        }
      }

      const config: ModalConfig<Props> = {
        id,
        component,
        props: passedProps,
        className: className || deprecatedCardClassName,
        label,
      }

      if (modals.isOpen(id)) {
        return modals.replace(id, config)
      }

      return modals.push(config)
    }) as UseModalResult<Props>

    modal.id = id
    modal.open = props => modal(props)
    modal.close = () => {
      return modals.close(id)
    }

    return modal
  }, [component, id, label])
}

export const useModals = (): ModalContextType => {
  const modalManager = useContext(ModalContext)
  if (!modalManager) {
    throw new Error(
      'ModalContext could not be consumed. Be sure to wrap the app with ModalProvider.',
    )
  }

  return modalManager
}
