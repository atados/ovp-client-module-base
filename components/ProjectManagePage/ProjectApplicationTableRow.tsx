import React from 'react'
import cx from 'classnames'
import { API } from '~/types/api'
import Avatar from '~/components/Avatar'
import LinkIf from '~/components/LinkIf'
import styled from 'styled-components'
import Icon from '../Icon'
import Tooltip from '../Tooltip'
import { isNotAppliedAnymore } from '~/lib/utils/project-application-utils'
import { Page, PageAs } from '~/common'
import ActivityIndicator from '../ActivityIndicator'

const AvatarStyled = styled(Avatar)`
  flex: 0 0 2.5rem;
  width: 2.5rem;
  height: 2.5rem;
`

interface ProjectApplicationTableRowProps {
  readonly application: API.ProjectApplication
  readonly className?: string
  readonly readOnly?: boolean
  readonly removing?: boolean
  readonly onConfirm?: (application: API.ProjectApplication) => void
  readonly onRemove?: (application: API.ProjectApplication) => void
}

const ProjectApplicationTableRow: React.FC<ProjectApplicationTableRowProps> = ({
  className,
  application,
  readOnly,
  onConfirm,
  onRemove,
  removing,
}) => {
  const { user, message, status, role } = application
  const name = user ? user.name : application.name
  const phone = user ? user.phone : application.phone
  const email = user ? user.email : application.email
  const contactColumnClassName = readOnly ? 'max-w-lg' : 'max-w-xs'

  return (
    <tr className={className}>
      <td className="px-4 py-3">
        <div className="flex">
          <LinkIf
            if={user}
            href={Page.PublicUser}
            as={PageAs.PublicUser({ slug: user && user.slug })}
          >
            <AvatarStyled
              image={user && user.avatar}
              className="rounded-full mr-3 bg-cover bg-center relative"
              fallBackClassName="bg-gray-300"
            >
              <span
                className={cx(
                  `absolute left-0 bottom-0 w-4 h-4 rounded-full z-20 border-2 border-white -ml-1 -mb-1`,
                  {
                    'bg-red-500': isNotAppliedAnymore(status),
                    'bg-yellow-500': status === 'applied',
                    'bg-green-500': status === 'confirmed-volunteer',
                  },
                )}
              />
            </AvatarStyled>
          </LinkIf>
          <div className="flex-grow">
            <LinkIf
              if={user}
              href={Page.PublicUser}
              as={PageAs.PublicUser({ slug: user && user.slug })}
            >
              <span className="text-md font-medium text-gray-800 block truncate">
                {name}
              </span>
            </LinkIf>
            {isNotAppliedAnymore(status) ? (
              <span className="text-sm block text-red-600 truncate ">
                {status === 'unapplied' ? 'Desistente' : 'Removido'}
              </span>
            ) : role ? (
              <span className="text-green-600 text-sm block truncate">
                {role.name}
              </span>
            ) : (
              <span className="text-gray-500 text-sm block truncate">
                Sem cargo definido
              </span>
            )}
            {message && (
              <span className="block text-sm text-gray-700 mt-1">
                {message}
              </span>
            )}
          </div>
        </div>
      </td>
      <td className={`${contactColumnClassName} px-4 py-3 truncate text-md`}>
        {phone && (
          <span
            className={`${contactColumnClassName} block text-sm text-gray-800 mt-1 truncate`}
          >
            <Icon name="phone" className="text-gray-600 mr-2" />
            {phone}
          </span>
        )}
        {email && (
          <a
            href={`mailto:${email}`}
            className={`${contactColumnClassName} block text-sm text-gray-800 mt-1 truncate hover:underline`}
          >
            <Icon name="email" className="text-gray-600 mr-2" />
            {email}
          </a>
        )}
      </td>
      {!readOnly ? (
        <td className="px-4 py-3 truncate text-lg" style={{ width: '200px' }}>
          <button
            type="button"
            className="btn text-white text-sm rounded-full bg-green-500 hover:bg-green-600"
            onClick={onConfirm ? () => onConfirm(application) : undefined}
            role="confirm-application"
          >
            <Icon name="check_circle" className="mr-1" />
            Confirmar inscrição
          </button>
          <Tooltip value="Remover inscrição">
            <button
              type="button"
              className="btn text-gray-700 text-sm rounded-full bg-gray-300 hover:bg-red-600 hover:text-white ml-2"
              onClick={onRemove ? () => onRemove(application) : undefined}
            >
              <Icon name="close" />
            </button>
          </Tooltip>
        </td>
      ) : onRemove && !readOnly ? (
        <td className="px-4 py-3 truncate text-lg" style={{ width: '80px' }}>
          <Tooltip value="Remover inscrição">
            <button
              type="button"
              className="btn text-gray-700 text-sm rounded-full bg-gray-300 hover:bg-red-600 hover:text-white ml-2"
              onClick={() => onRemove(application)}
            >
              {removing ? (
                <ActivityIndicator fill="#333" size={24} />
              ) : (
                <Icon name="close" />
              )}
            </button>
          </Tooltip>
        </td>
      ) : (
        <td></td>
      )}
    </tr>
  )
}

ProjectApplicationTableRow.displayName = 'ProjectApplicationTableRow'

export default ProjectApplicationTableRow
