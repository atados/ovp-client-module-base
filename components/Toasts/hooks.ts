import { useRef, useEffect } from 'react'
import { useToasts } from '~/components/Toasts'

export const notifyErrorWithToast = (error?: Error | null) => {
  const toasts = useToasts()
  const toastIdRef = useRef<string | null>(null)

  useEffect(() => {
    if (toastIdRef.current && !toasts.get(toastIdRef.current)) {
      toastIdRef.current = null
    }

    if (!error) {
      if (toastIdRef.current) {
        toasts.remove(toastIdRef.current)
        toastIdRef.current = null
      }

      return
    }
    if (toastIdRef.current) {
      toasts.replace(toastIdRef.current, error.message)
      return
    }

    toastIdRef.current = toasts.add(error.message, 'error')
  }, [error])

  return toastIdRef.current
}
