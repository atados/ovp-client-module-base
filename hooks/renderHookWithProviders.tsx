import { renderHook, RenderHookOptions } from '@testing-library/react-hooks'
import { Provider } from 'react-redux'
import { Store } from 'redux'

import { configureStore } from '~/redux/with-redux'

interface CreateProvidersWrapperConfig {
  store: Store
}

const createProvidersWrapper = ({ store }: CreateProvidersWrapperConfig) => ({
  children,
}) => {
  return <Provider store={store}>{children}</Provider>
}

const renderHookWithProviders = <P, R>(
  fn: (props: P) => R,
  {
    store = configureStore({ startup: undefined }),
    ...options
  }: Partial<CreateProvidersWrapperConfig> & RenderHookOptions<P> = {},
) =>
  renderHook<P, R>(fn, {
    ...options,
    wrapper: createProvidersWrapper({ store }),
  })

export default renderHookWithProviders
