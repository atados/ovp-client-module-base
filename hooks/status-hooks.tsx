import { useEffect, useContext, useState, useMemo, useCallback } from 'react'
import { StatusLevel } from '../components/Status'
import StatusContext from '../components/Status/StatusContext'
import { reportError } from '../lib/utils/error'

interface Message {
  id: string
  text: string
  level: StatusLevel
}
export function useSetStatus() {
  const context = useContext(StatusContext)
  const [message, setMessage] = useState<Message | null>(null)

  const setStatus = useMemo(() => {
    function fn(text: Error): void
    function fn(text: string, level: StatusLevel): void
    function fn(text: string | Error, level?: StatusLevel): void {
      if (!context) {
        throw new Error('Status context not found')
      }

      if (text instanceof Error) {
        reportError(text)
        setMessage({
          id: context.register(text.message, StatusLevel.Error),
          text: text.message,
          level: StatusLevel.Error,
        })
        return
      }

      if (level) {
        setMessage({ id: context.register(text, level), text, level })
      }
    }

    return fn
  }, [context])
  const removeStatusIfExists = useCallback(() => {
    if (!context) {
      throw new Error('Status context not found')
    }

    if (message) {
      context.unRegister(message.id)
    }
  }, [context, message])

  useEffect(() => () => removeStatusIfExists(), [message])
  return useMemo(() => [setStatus, removeStatusIfExists], [setMessage])
}
