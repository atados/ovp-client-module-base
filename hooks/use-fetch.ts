import useSWR, { ConfigInterface } from 'swr'
import { dev } from '~/common/constants'

export interface UseFetchOptions<Data, Error>
  extends ConfigInterface<Data, Error> {
  context?: RequestInit
}

/**
 *
 * @param urlOrURLCreator
 * @param options
 * @example const { data, error, isValidating } = useFetch('http://google.com/api/json')
 */
export const useFetch = <Data = any, Error = any>(
  urlOrURLCreator: string | (() => string),
  options: UseFetchOptions<Data, Error> = {},
) => {
  const result = useSWR<Data>(
    urlOrURLCreator,
    url => {
      if (url === null) {
        return Promise.resolve()
      }

      return swrFetcher(url, options.context)
    },
    options,
  )

  return result
}

const logFetcher = (method: string = 'GET', ...args: any[]) =>
  console.log(
    `%c [FETCH] ${method}`,
    'color: green; font-size: 12px; font-weight: bolder',
    ...args,
  )
export const swrFetcher = (url: string, options?: RequestInit) => {
  if (dev) {
    logFetcher(options?.method, url)
  }

  return fetch(url, options).then(res => {
    const data = res.json()
    if (dev) {
      logFetcher(options?.method, url, data)
    }

    return data
  })
}
