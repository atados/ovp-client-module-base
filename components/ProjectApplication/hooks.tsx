import { useSelector } from 'react-redux'
import { RootState } from '~/redux/root-reducer'
import { reportError } from '~/lib/utils/error'
import { Project } from '~/redux/ducks/project'
import { useModal } from '~/components/Modal'
import ProjectApplication, {
  ProjectApplicationProps,
} from '~/components/ProjectApplication/ProjectApplication'
import ProjectApplicationRegistry from '~/components/ProjectApplication/ProjectApplicationRegistry'
import CompleteViewerProfileForm from '~/components/CompleteViewerProfileForm'
import useModalManager from '~/hooks/use-modal-manager'
import Authentication from '~/components/Authentication'
import ConfirmEmailBeforeProceeding from '~/components/EmailConfirmation/ConfirmEmailBeforeProceeding'
import { Config } from '~/common'

export function useProjectApplication() {
  const viewer = useSelector((state: RootState) => state.user)
  const modalManager = useModalManager()
  const openApplicationRegistryModal = useModal({
    id: 'ProjectApplicationRegistry',
    component: ProjectApplicationRegistry,
    cardClassName: 'p-4',
  })
  const openApplicationModal = useModal<ProjectApplicationProps>({
    id: 'ProjectApplication',
    component: ProjectApplication,
  })
  const openCompleteViewerInformationModal = useModal({
    id: 'CompleteViewerProfileForm',
    component: CompleteViewerProfileForm,
  })
  const openConfirmEmailBeforeProceedingModal = useModal({
    id: 'ConfirmEmailBeforeProceeding',
    component: ConfirmEmailBeforeProceeding,
  })
  const openAuthenticationModal = useModal({
    id: 'Authentication',
    component: Authentication,
    cardClassName: 'p-4',
  })

  return (project: Project, defaultRoleId?: number) => {
    const onApply = () =>
      openApplicationModal({
        project,
        defaultRoleId,
        next: () => {
          modalManager.close('ProjectApplication')
          openApplicationRegistryModal({ project, defaultRoleId })
        },
      })

    if (!viewer) {
      openAuthenticationModal({
        onAuthenticate: () => {
          modalManager.close('Authentication')
          onApply()
        },
      })
      return
    }

    if (
      !viewer.is_email_verified &&
      Config.application.require.emailConfirmation
    ) {
      openConfirmEmailBeforeProceedingModal()
      return
    }

    if (project.current_user_is_applied) {
      openApplicationRegistryModal({ project, defaultRoleId })
      return
    }

    if (viewer.profile) {
      if (
        !viewer.phone ||
        !viewer.profile ||
        !viewer.profile.birthday_date ||
        !viewer.profile.gender ||
        !viewer.profile.address
      ) {
        openCompleteViewerInformationModal({
          next: onApply,
        })
      } else {
        onApply()
      }
    } else {
      reportError(
        `Viewer Profile is ${JSON.stringify(
          viewer,
        )} at useProjectApplication. Ignored user info.`,
      )
      onApply()
    }
  }
}
