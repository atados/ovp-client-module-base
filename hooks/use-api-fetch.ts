import { UseFetchOptions, useFetch } from '~/hooks/use-fetch2'
import { API_URL } from '~/common/constants'

/**
 *
 * @param endpointOrEndpointCreator The API endpoint to fetch from
 * @param options Options passed to SWR
 * @example const { data, error, loading} = useFetchAPI('/search/projects')
 */
export const useAPIFetch = <Data = any, Error = any>(
  endpointOrEndpointCreator: string | (() => string),
  options?: UseFetchOptions<Data, Error>,
) => {
  return useFetch(() => {
    if (typeof endpointOrEndpointCreator === 'function') {
      return `${API_URL}${endpointOrEndpointCreator()}`
    }

    return `${API_URL}${endpointOrEndpointCreator}`
  }, options)
}
