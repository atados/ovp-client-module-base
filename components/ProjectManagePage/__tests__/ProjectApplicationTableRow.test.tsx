import ProjectApplicationTableRow from '../ProjectApplicationTableRow'
import { render, fireEvent, wait } from '@testing-library/react'
import * as legacyFetchHooks from '~/hooks/use-fetch'
import renderer from 'react-test-renderer'

jest.mock('isomorphic-unfetch', () => jest.fn())
jest.mock('~/hooks/use-fetch', () => ({
  useAPIFetcher: jest.fn().mockImplementation(() => ({
    fetch: jest.fn(),
  })),
  mutateFetchCache: jest.fn(),
}))

const useAPIFetcher = (legacyFetchHooks.useAPIFetcher as any) as jest.Mock<
  typeof legacyFetchHooks['useAPIFetcher']
>
describe('ProjectApplicationTableRow', () => {
  beforeEach(() => useAPIFetcher.mockClear())

  it('should match snapshot', () => {
    const tree = renderer
      .create(
        <table>
          <tbody>
            <ProjectApplicationTableRow
              projectSlug="slug"
              application={{
                status: 'applied',
                canceled: false,
                date: String(Date.now()),
                id: 1,
              }}
            />
          </tbody>
        </table>,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should confirm an application when the confirm button is clicked', async () => {
    const handleConfirm = jest.fn()
    const { getByRole } = render(
      <table>
        <tbody>
          <ProjectApplicationTableRow
            projectSlug="slug"
            application={{
              status: 'applied',
              canceled: false,
              date: String(Date.now()),
              id: 1,
            }}
            onConfirm={handleConfirm}
          />
        </tbody>
      </table>,
    )

    const [{ value: fetchHook }] = useAPIFetcher.mock.results
    fireEvent.click(getByRole('confirm-application'))
    await wait()
    const [actionCreator] = useAPIFetcher.mock.calls[0]
    expect(actionCreator('confirmed-volunteer')).toMatchObject({
      method: 'PATCH',
      endpoint: '/projects/slug/applies/1/',
      meta: {
        to: 'confirmed-volunteer',
      },
    })
    expect(fetchHook.fetch).toBeCalledWith('confirmed-volunteer')
    expect(handleConfirm).toBeCalledTimes(1)
  })

  it('should not allow an application to be confirmed when it already is', async () => {
    const handleConfirm = jest.fn()
    const { queryByRole } = render(
      <table>
        <tbody>
          <ProjectApplicationTableRow
            projectSlug="slug"
            application={{
              status: 'confirmed-volunteer',
              canceled: false,
              date: String(Date.now()),
              id: 1,
            }}
            onConfirm={handleConfirm}
          />
        </tbody>
      </table>,
    )

    const [{ value: fetchHook }] = useAPIFetcher.mock.results
    expect(queryByRole('confirm-application')).toBeNull()
    expect(fetchHook.fetch).toBeCalledTimes(0)
    expect(handleConfirm).toBeCalledTimes(0)
  })

  it('should remove an application if the button is clicked', async () => {
    const handleRemoval = jest.fn()
    const { getByRole } = render(
      <table>
        <tbody>
          <ProjectApplicationTableRow
            projectSlug="slug"
            application={{
              status: 'confirmed-volunteer',
              canceled: false,
              date: String(Date.now()),
              id: 1,
            }}
            onRemove={handleRemoval}
          />
        </tbody>
      </table>,
    )

    const [{ value: fetchHook }] = useAPIFetcher.mock.results
    fireEvent.click(getByRole('remove-application'))
    await wait()
    const [actionCreator] = useAPIFetcher.mock.calls[0]
    expect(actionCreator('unapplied-by-deactivation')).toMatchObject({
      method: 'PATCH',
      endpoint: '/projects/slug/applies/1/',
      meta: {
        to: 'unapplied-by-deactivation',
      },
    })
    expect(fetchHook.fetch).toBeCalledWith('unapplied-by-deactivation')
    expect(handleRemoval).toBeCalledTimes(1)
  })
})
