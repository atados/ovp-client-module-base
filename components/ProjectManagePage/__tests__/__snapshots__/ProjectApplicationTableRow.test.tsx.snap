// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectApplicationTableRow should match snapshot 1`] = `
.c0 {
  background-position: center;
  background-repeat: no-repeat;
}

.c1 {
  -webkit-flex: 0 0 2.5rem;
  -ms-flex: 0 0 2.5rem;
  flex: 0 0 2.5rem;
  width: 2.5rem;
  height: 2.5rem;
}

<table>
  <tbody>
    <tr>
      <td
        className="px-4 py-3"
      >
        <div
          className="flex"
        >
          <div
            className="c0 c1 rounded-full mr-3 bg-cover bg-center relative bg-gray-300"
          >
            <span
              className="absolute left-0 bottom-0 w-4 h-4 rounded-full z-20 border-2 border-white -ml-1 -mb-1 bg-yellow-500"
            />
          </div>
          <div
            className="flex-grow"
          >
            <span
              className="text-md font-medium text-gray-800 block truncate"
            />
            <span
              className="text-gray-500 text-sm block truncate"
            >
              FormattedMessage
            </span>
          </div>
        </div>
      </td>
      <td
        className="max-w-xs px-4 py-3 truncate text-md"
      />
      <td
        className="px-4 py-3 truncate text-lg"
        style={
          Object {
            "width": "200px",
          }
        }
      >
        <button
          className="btn text-white text-sm rounded-full bg-green-500 hover:bg-green-600"
          onClick={[Function]}
          role="confirm-application"
          type="button"
        >
          <span
            className="icon im mr-1"
          >
            check_circle
          </span>
          FormattedMessage
        </button>
        <button
          className="btn text-gray-700 text-sm rounded-full bg-gray-300 hover:bg-red-600 hover:text-white ml-2"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          role="remove-application"
          type="button"
        >
          <span
            className="icon im"
          >
            close
          </span>
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;
